text/0477-blueprints-update.md

```diff
@@ -0,0 +1,248 @@
+    "blueprints": [
+      {
+        "name": "my-custom-blueprint",
+        "version": "0.0.1"
+      }
+    ]
+  }
+}
+```
+
+I vote “ember-cli-update.json”. Since the code will be modified by the updater, it’s easier to reformat JSON. “.ember-cli” I think supports JS, so it would be hard to modify. “package.json” seems like an abuse to put more metadata in there.
+
+I've found via further testing that mixing a file that is tracked by the default ember-cli blueprint (.ember-cli, package.json) and modified by ember-cli-update (to update the blueprint metadata), it gets cumbersome. Both processes try to edit the same file. So in theory, ember-cli could alter the .ember-cli file in a way that conflicts with what you have in yours. This would give you a git conflict, then the updater can't read it anymore because it is in a state of invalid JS. I think this further supports a separate file.
+
+How should the state be structured?
+
+```js
+{
+  "blueprints": [
+    {
+      "name": "my-custom-blueprint",
+      "version": "0.0.1"
+    }
+  ]
+}
+```
+
+```js
+{
+  "blueprints": {
+    "my-custom-blueprint": {
+      "version": "0.0.1"
+    }
+  }
+}
+```
+
+Future-proofing for more options
+
+```js
+{
+  "additionalFutureOptions": null,
+  "blueprints": {
+    "my-custom-blueprint": {
+      "version": "0.0.1",
+      "additionalFutureOptions": null,
+    }
+  }
+}
+```
+
+**Methods of delivery**
+
+Blueprints could be responsible for injecting the state into projects
+
+```js
+// my-custom-blueprint/blueprints/my-custom-blueprint/files/.ember-cli
+{
+  "ember-cli-update": {
+    "blueprints": [
+      {
+        "name": "my-custom-blueprint",
+        "version": "<%= blueprintVersion %>"
+      }
+    ]
+  }
+}
+```
+
+or ember-cli-update could manage it itself if the blueprint does not provide this info. You would have to provide the initial blueprint version to start from.
+
+The goal is for when you type `ember-cli-update` in your project, you will get a prompt like
+
+```
+$ ember-cli-update
+Multiple blueprint updates found, which would you like to update?
+ * ember-cli
+ * ember-cli-mirage
+ * ember-ci
+```
+
+The reason this is so powerful is any org could create their own _partial_ project blueprint (public or private). This blueprint can make any slight (or massive) tweaks to the official blueprints, and ember-cli-update can keep any project in sync with both the official blueprint and your org's blueprints, at the same time.
+
+**Complete vs partial**
+
+We need a way to denote a project replacement blueprint and a supplemental blueprint. A project replacement removes any sort of default ember-cli blueprint tracking, even if it still looks like a normal Ember project (has ember-cli, ember-addon, etc.). A supplemental blueprint is one that piggy-backs on a complete blueprint and makes certain tweaks to files. Think of ember-cli-mirage's extra project configs, or your company's custom settings on top of a normal Ember app.
```

Comment:
Why do we need to know the difference between complete and partial? I think you are basically saying that we need to know the difference between the _original_ project blueprint (aka the "root" blueprint) and the others so that we know how to build up the comparison diff, is that right?

Reply:
At the time of writing this, I thought I would need to know the difference between the two to do the correct thing, but after implementing most the functionality, I don't think that's true anymore.